### DISCUSSION TITLE
Batch crop to closest defied ratio

### DISCUSSION THREAD

**Discussion by HameedFarah** [ImageMagick/ImageMagick#5235] [Digital Image Processing]
Hello great IMers :)

I am using the latest version of ImageMagick on Windows 10
I have over 2000 photos that I want to crop to certain ration depending on the closest defined ratio.

The ratios I need are 1:1 and 7:5 and 1:3 They can be both portrait and landscape

Examples,
Original dimensions (px) => Cropped dimensions (px)
100:110 => 100:100
159:50 => 150:50
42:120 => 40:120

How can I do that?

*Bonus point :) I am not sure if IM can do this, but if it can do "Smart" cropping and detect the important part of the image that would be great

Thank you

**Comment by snibgo** [comment]
> *Bonus point :) I am not sure if IM can do this, but if it can do "Smart" cropping and detect the important part of the image that would be great

I show how this can be done at [Crop to detail](http://im.snibgo.com/crop2det.htm). That includes Windows BAT scripts.

On the aspect ratios: interesting problem (which means I don't have an immediate solution).

I suggest making a list of w/h for each defined ratio, including portrait and landscape separately. So we have: 1, 1.4, 0.33333, 0.71429, 3.

For each image calculate w/h. If that is equal to one of the defined ratios, then that is the ratio to use. But if the image is 159x50, then w/h=3.18, which isn't an exact defined ratio. Which is the closest?

We might use the absolute difference between the image aspect ratio and the defined ratios. The smallest difference is the one we want. For example:
```
abs(3.18 - 1) = 2.18
abs(3.18 - 1.4) = 1.78
abs(3.18 - 0.33333) = 2.84667
abs(3.18 - 0.71429) = 2.46571
abs(3.18 - 3) = 0.18
```
The smallest of these is 0.18, so the closest defined ratio is "3".

However, I have a feeling that this definition of "closest" isn't technically the best one. It will work, but a better definition might be to square the ratios before taking the difference. Or we should divide the ratios instead of taking the difference. Or something.

**Reply by HameedFarah** [reply]
Thank you for your help.
I like your thinking!

To make it more interesting, is there a way to combine all of these?
Meaning for IM to first choose the best crop ratio and dimensions then crop the image based on the "Crop to Detail"?

**Comment by snibgo** [comment]
Yes, techniques can be combined.

My script cropToDetailSml.bat takes required dimensions in pixels, rather than an aspect ratio. The script can be modified, of course, but then the problem is under-specified. It works by first creating a saliency map, a grayscale version of the input where intensity represents the amount of detail. (A different saliency map might come from searches for human faces, or a specific colour, and so on.)

Suppose we have an image and its saliency map that are both 2000x3000 pixels, and we want an image with aspect ratio 2:3. The result might be 2000x3000 or 1000x1500 or 200x300 or 20x30 or 2x3 or many other possibilities. How is an algorithm to choose between them? It might choose the size that has the maximum average saliency, but this will mostly (always?) choose the smallest size. For most purposes, a 2x3 image is useless. We might add a simple rule like "don't make an image smaller than 600x600", but that will mostly (always?) choose the size with aspect ratio 2:3 that fits exactly in that box, ie 400x600 pixels.

Instead of measuring "maximum average saliency", we might measure how much saliency that is present in the input image is captured by a given crop. (Perhaps the calculation is simply: average_intensity \* height \* width.) This would encourage trimming off areas with low saliency (little detail), and discourage trimming off areas with high saliency. I think that could be made to work. But it might be slow.

And when that works, an overall system could figure out the best of the defined aspect ratios, based on how much saliency is captured, instead of being based on the aspect ratio of the input image.

**Reply by HameedFarah** [reply]
I am looking for the largest possible resulting image that has one of the required aspect ratios. So no need to do any "saliency mapping". 

So I guess it would be something like this,
Determine the current image aspect ratio
Find the closest aspect ratio from the provided options
Use the original image dimension and multiply it by the ratio to find the output image dimensions
Use the CropToDetail to find the best location to center the crop

Added steps would be to input the required folder with the images, then the script would output the cropped images to a subfolder.

So in case 2000x3000 and the ratio is 2:3 then no cropping is needed.
If it was 2050x3011 so the closest ratio would be 2:3. The result crop size would be 2007x3011

**Comment by snibgo** [comment]
Okay, you want the crop to do nothing, or to reduce height, or to reduce width, and never to reduce both height and width. Is that correct?

Here is a script I call "nearestAspect.bat". Most of it is to find the aspect ratio that removes the fewest pixels. The list of possible ARs is defined neat the top. Then it calls cropToDetailSml.bat to find the crop offsets that give the most detail. That script creates and uses a saliency map.
```
rem From image %1,
rem find the nearest aspect ratio from a list.
rem Writes output cropped image to %2.
rem "Nearest" means the AR that results in a crop removing the fewest pixels.

setlocal enabledelayedexpansion

set INFILE=%1
set OUTFILE=%2

if "%OUTFILE%"=="" (
  echo %0: needs two arguments, for infile and outfile
  exit /B 1
)

set ARLIST=1:1,7:5,1:3,5:7,3:1,2:3,3:2

set SML_DIFF=99999999
set SML_AR=

for %%A in (%ARLIST%) do (

  echo AR is %%A

  for /F "usebackq" %%L in (`%IMG7%magick ^
    -precision 15 ^
    %INFILE% ^
    -format "inpt=%%[fx:w*h]\n" ^
    +write info: ^
    -crop %%A+0+0 +repage ^
    -format "crpd=%%[fx:w*h]\nWW=%%w\nHH=%%h\n" ^
    info:`) do set %%L

  set /A diff=!inpt!-!crpd!

  if !SML_DIFF! GTR !diff! (
    set SML_DIFF=!diff!
    set SML_AR=%%A
    set SML_WW=!WW!
    set SML_HH=!HH!
  )
  echo diff=!diff!
)

if "%SML_AR%"=="" (
  echo %0: Something bad happened.
  exit /B 1
)

echo SML_DIFF=%SML_DIFF% SML_AR=%SML_AR% SML_WW=%SML_WW% SML_HH=%SML_HH%

call %PICTBAT%cropToDetailSml.bat %INFILE% %OUTFILE% %SML_WW%x%SML_HH%

endlocal
```
Call the script like this:
```
call %PICTBAT%nearestAspect toes.png x.png
```
All my image-processing scripts reside in the directory "PICTBAT".

This finds the best crop for toes.png, and writes the output to x.png.

![x](https://user-images.githubusercontent.com/33812368/174399009-c7a73cdd-fe39-4535-87cf-a96164923340.png)

**Reply by HameedFarah** [reply]
Thank you so much for your reply but running nearestAspect.bat gives a couple of errors. I have attached the output

FYI I have IM in a folder in D: and the test image is called a.jpg

I also created PICTBAT directory and placed the two bat files in it but that didn't work so they are now located in IM folder along with the jpg test image.

**Comment by snibgo** [comment]
> I have attached the output

Did you?

cropToDetailSml.bat calls other scripts. Perhaps you haven't downloaded those?

**Reply by HameedFarah** [reply]
It seems you are correct as usual :)
The script is calling some other scripts which are not found and I cannot find them on the page like 'echoOffSave'

And here is the output file this time
[output.txt](https://github.com/ImageMagick/ImageMagick/files/8931810/output.txt)

**Comment by snibgo** [comment]
I suggest you visit [Zipped BAT files](http://im.snibgo.com/zipbats.htm). Near the top is a link to a zip file containing many of my scripts. Unzip them to your directory of choice.

Incidentally, when I have...
```
call %PICTBAT%nearestAspect toes.png x.png
```
... this means PICTBAT is an environment variable that points to a directory.
